Allow major crates to opt-out of reflection

ManevilleF opened the pull request:
# Objective

I use the `bevy_ecs` without `bevy_app` as I need a more custom solution. Some very useful crates like `bevy_time` for Time running criteria and `bevy_hierarchy` all provide a `bevy_app::Plugin` and have extra dependencies I don't need like `bevy_reflect` when I can disable it directly in `bevy_ecs`

## Solution

I added feature gates for `bevy_app`, the reflection and extra features for a more "opt-in" behaviour for people using mainly the ECS

DJMcNab commented:
I'm not sure I like this. Having an opt-out for reflection makes some amount of sense, but I'm less convinced by every other change.

We cannot reasonably test that things compile with all combinations of features, and this adds a lot of complexity for exactly you. This also appears to be only for reducing compile time at that, not any fundamental change in capabilities - you can just not call `Plugin::build` if you don't want to use it, and it should be optimised out by the linker, including therefore all the reflection code.

If you still want this, you should probably just vendor the dependencies yourself and trim them that way - it's not like you'll be using any third party crates with this, since they'll all be correctly using the `Plugin` abstraction anyway. [`cargo vendor`](https://doc.rust-lang.org/cargo/commands/cargo-vendor.html) is the tool for this.

ManevilleF commented:
I removed some extra changes, leaving mainly the opt-out for `bevy_reflect`, which I think should be applied to most crates.
I also kept the `crossbeam-channel` opt-out for `bevy_time` which is used specifically for `bevy_render`.

I doubt I'm the only person needing some hierarchy and fixed time steps while not needing reflection. `bevy_ecs` itself has an opt out feature for reflection, and is meant to work on standalone. I am bringing the same concept to other core crates

cart commented:
Yeah I'm pretty sure I'm also against this. It affects the legibility of core types and the complexity of the code / config for optimizing a corner case. Bevy ECS is indeed intended to be usable standalone, but I think the buck should stop there. Core Bevy functionality crates should be able to assume that core pieces like reflection are available.

We can leave this open for a bit to let other people weigh in, but I'm inclined to close this.

cart commented:
I'm closing this as my opinion on this is pretty firm, but feel free to keep the discussion going if you have counter arguments.
